lib/puppet/pops/types/p_encrypted_type.rb

```diff
@@ -0,0 +1,354 @@
+
+      cipher = best_matching_cipher(ciphers)
+      if cipher.nil?
+        if ciphers.is_a?(Array)
+          raise ArgumentError, _("None of the cipher names \"%{ciphers}\" are supported. Supported ciphers: %{available_ciphers}") % {
+            available_ciphers: acceptable_and_available_ciphers,
+            ciphers: ciphers
+          }
+        else
+          raise ArgumentError, _("Unsupported cipher algorithm \"%{cipher_name}\". Supported ciphers: %{available_ciphers}") % {
+            available_ciphers: acceptable_and_available_ciphers,
+            cipher_name: cipher
+          }
+        end
+      end
+
+      key = certificate.content.public_key
+      fingerprint = certificate.fingerprint
+      data = serialize(data)
+      format = 'json,' + cipher
+
+      aes_encrypt = OpenSSL::Cipher.new(cipher).encrypt
+
+      # Use a random key
+      aes_encrypt.key = aes_key = aes_encrypt.random_key
+
+      # Use a random initialization vector (safer) - these 16 bytes are prepended to the clear text
+      # and dropped after decryption.
+      #
+      iv = aes_encrypt.random_iv
+
+      # Encrypt the data with this key
+      crypt = iv + aes_encrypt.update(data) << aes_encrypt.final
+      # Encrypt the random key with the public key
+      encrypted_key = rsa_key(key).public_encrypt(aes_key)
+
+      # Encrypt the fingerprint with public key
+      aes_encrypt.reset
+
+      # use different iv
+      iv = aes_encrypt.random_iv
+      encrypted_fingerprint = iv + aes_encrypt.update(fingerprint) << aes_encrypt.final
+
+      Encrypted.new(format, encrypted_key, crypt, encrypted_fingerprint)
+    end
+
+    # @api private
+    def self.rsa_key(key)
+      OpenSSL::PKey::RSA.new(key)
+    end
+    private_class_method :rsa_key
+
+    # @api private
+    def self.serialize(data)
+      io = StringIO.new
+      writer = Puppet::Pops::Serialization::JSON::Writer.new(io)
+      serializer = Puppet::Pops::Serialization::Serializer.new(writer)
+      serializer.write(data)
+      serializer.finish
+      io.string
+    end
+    private_class_method :serialize
+
+    # Returns the union of acceptable and available ciphers
+    def self.acceptable_and_available_ciphers
+      # return from cache unless accepted_ciphers changed since last cached
+      return @supported_ciphers if !@supported_ciphers.nil? && @supported_ciphers_id == Puppet[:accepted_ciphers].__id__
+      available = OpenSSL::Cipher.ciphers
+      acceptable = Puppet[:accepted_ciphers]
+      @supported_ciphers_id = Puppet[:accepted_ciphers].__id__
+      @supported_ciphers = acceptable.empty? ? available : acceptable & available
+      @supported_ciphers
+    end
+
+    def self.best_matching_cipher(ciphers)
+      if ciphers.nil?
+        # use the default - first accepted, or AES-256-CBC if accepted is not defined
+        accepted = Puppet[:accepted_ciphers]
+        ciphers = [accepted.empty? ? 'AES-256-CBC' : accepted[0]]
```

Comment:
Probably worth using GCM mode by default, so that we can provide integrity protection of the encrypted values, to protect against bitflipping.